packages/fxa-auth-server/lib/routes/subscriptions.js

```diff
@@ -10,6 +10,597 @@ const ScopeSet = require('../../../fxa-shared').oauth.scopes;
+        throw error.unknownSubscription();
+      }
+    }
+
+    const subscription = await this.payments.stripe.subscriptions.update(
+      subscriptionId,
+      {
+        cancel_at_period_end: false,
+      }
+    );
+    if (!['active', 'trialing'].includes(subscription.status)) {
+      const err = new Error(
+        `Reactivated subscription (${subscriptionId}) is not active/trialing`
+      );
+      throw error.backendServiceFailure(
+        'stripe',
+        'reactivateSubscription',
+        {},
+        err
+      );
+    }
+
+    await this.db.reactivateAccountSubscription(uid, subscriptionId);
+
+    await this.push.notifyProfileUpdated(uid, await request.app.devices);
+    this.log.notifyAttachedServices('profileDataChanged', request, {
+      uid,
+      email,
+    });
+    await this.profile.deleteCache(uid);
+
+    this.log.info('subscriptions.reactivateSubscription.success', {
+      uid,
+      subscriptionId,
+    });
+
+    return {};
+  }
+
+  async updateSubscription(request) {
+    this.log.begin('subscriptions.updateSubscription', request);
+
+    const { uid, email } = await handleAuth(this.db, request.auth, true);
+
+    await this.customs.check(request, email, 'updateSubscription');
+
+    const { subscriptionId } = request.params;
+    const { planId } = request.payload;
+
+    let accountSub;
+
+    try {
+      accountSub = await this.db.getAccountSubscription(uid, subscriptionId);
+    } catch (err) {
+      if (err.statusCode === 404 && err.errno === 116) {
+        throw error.unknownSubscription();
+      }
+    }
+    const oldProductId = accountSub.productId;
+
+    // Verify the plan is a valid upgrade for this subscription.
+    await this.payments.verifyPlanUpgradeForSubscription(oldProductId, planId);
+
+    // Upgrade the plan
+    const changeResponse = await this.payments.changeSubscriptionPlan(
+      subscriptionId,
+      planId
+    );
+    const newProductId = changeResponse.plan.product;
+
+    // Update the local db record for the new plan. We don't have a method to
+    // change the product on file for a sub thus the delete/create here even
+    // though its more work to catch both errors for a retry.
+    try {
+      await this.db.deleteAccountSubscription(uid, subscriptionId);
+    } catch (err) {
+      // It's ok if it was already cancelled or deleted.
+      if (err.statusCode !== 404) {
+        throw err;
+      }
+    }
+
+    // This call needs to succeed for us to consider this a success.
+    await this.db.createAccountSubscription({
+      uid,
+      subscriptionId,
+      productId: newProductId,
+      createdAt: Date.now(),
+    });
+
+    const devices = await request.app.devices;
+    await this.push.notifyProfileUpdated(uid, devices);
+    this.log.notifyAttachedServices('profileDataChanged', request, {
+      uid,
+      email,
+    });
+    await this.profile.deleteCache(uid);
+
+    return { subscriptionId };
+  }
```

Comment:
Scrolling through these *Subscription functions, it looks like there's quite a bit of common code.  I wonder if we could write a function to create these functions.  Not sure if it's worthwhile though.

Reply:
Once production goes to direct stripe, we're going to remove the other version entirely.

Reply:
*Some* of this will be rolled up into a common function in my branch that's a follow-on to this PR. Namely, the bits that we have to do when a customer / profile changes:

https://github.com/mozilla/fxa/compare/feat/issue-3809...lmorchard:3808-customer-caching#diff-eedc9e44d94c523680530e9c145c9496R62